src/utils.jl

```diff
@@ -123,23 +123,12 @@ end
 
 # Handling ComponentArrays
 # NOTE(@avik-pal): We should probably upsteam some of these
-function Base.zero(c::ComponentArray{T, N, <:CuArray{T}}) where {T, N}
-    return ComponentArray(zero(getdata(c)), getaxes(c))
-end
-
 Base.vec(c::ComponentArray) = getdata(c)
 
-Base.:-(x::ComponentArray) = ComponentArray(-getdata(x), getaxes(x))
-
 function Base.similar(c::ComponentArray, l::Vararg{Union{Integer, AbstractUnitRange}})
     return similar(getdata(c), l)
 end
 
-function Functors.functor(::Type{<:ComponentArray}, c)
```

Comment:
This function should not be removed.

Reply:
What's the reason? It would make `Optimisers` error.
```julia
julia> ps = ComponentArray(cu[1f0,2f0,3f0,4f0], Axis(a=1:2, b=3:4));
julia> st = Optimisers.setup(opt, ps)
(a = Leaf(Adam{Float32}(0.001, (0.9, 0.999), 1.19209f-7), (Float32[0.0, 0.0], Float32[0.0, 0.0], (0.9, 0.999))), b = Leaf(Adam{Float32}(0.001, (0.9, 0.999), 1.19209f-7), (Float32[0.0, 0.0], Float32[0.0, 0.0], (0.9, 0.999))))
julia> new_st, new_ps = Optimisers.update(st, ps, ps)
ERROR: CuArray only supports element types that are stored inline
Stacktrace:
  [1] error(s::String)
    @ Base .\error.jl:33
  [2] CuArray{Vector{Any}, 1, CUDA.Mem.DeviceBuffer}(#unused#::UndefInitializer, dims::Tuple{Int64})
    @ CUDA C:\Users\Luffy\.julia\packages\CUDA\DfvRa\src\array.jl:34
  [3] similar(a::CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, #unused#::Type{Vector{Any}}, dims::Tuple{Int64})
    @ CUDA C:\Users\Luffy\.julia\packages\CUDA\DfvRa\src\array.jl:166
  [4] _mapreduce(f::typeof(identity), op::ComponentArrays.var"#28#29", As::CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}; dims::Colon, init::Vector{Any})
    @ GPUArrays C:\Users\Luffy\.julia\packages\GPUArrays\Hyss4\src\host\mapreduce.jl:64
  [5] mapreduce(::Function, ::Function, ::CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}; dims::Function, init::Vector{Any})
    @ GPUArrays C:\Users\Luffy\.julia\packages\GPUArrays\Hyss4\src\host\mapreduce.jl:31
  [6] reduce(op::Function, A::CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}; kw::Base.Pairs{Symbol, Vector{Any}, Tuple{Symbol}, NamedTuple{(:init,), Tuple{Vector{Any}}}})
    @ Base .\reducedim.jl:371
  [7] pushcat!(a::Vector{Any}, b::CuArray{Float32, 1, CUDA.Mem.DeviceBuffer})
    @ ComponentArrays C:\Users\Luffy\.julia\packages\ComponentArrays\Eoni9\src\componentarray.jl:223
  [8] make_idx(data::Vector{Any}, x::CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, last_val::Int64)
    @ ComponentArrays C:\Users\Luffy\.julia\packages\ComponentArrays\Eoni9\src\componentarray.jl:178
  [9] make_idx(data::Vector{Any}, nt::NamedTuple{(:a, :b), Tuple{CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}}}, last_val::Int64)
    @ ComponentArrays C:\Users\Luffy\.julia\packages\ComponentArrays\Eoni9\src\componentarray.jl:156
 [10] make_carray_args(A::Type{Vector}, nt::NamedTuple{(:a, :b), Tuple{CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}}})
    @ ComponentArrays C:\Users\Luffy\.julia\packages\ComponentArrays\Eoni9\src\componentarray.jl:146
 [11] make_carray_args(nt::NamedTuple{(:a, :b), Tuple{CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}}})
    @ Lux d:\GitHub\Lux.jl\src\utils.jl:138
 [12] ComponentArray(nt::NamedTuple{(:a, :b), Tuple{CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}}})
    @ ComponentArrays C:\Users\Luffy\.julia\packages\ComponentArrays\Eoni9\src\componentarray.jl:63
 [13] update!(tree::NamedTuple{(:a, :b), Tuple{Optimisers.Leaf{Adam{Float32}, Tuple{CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, Tuple{Float32, Float32}}}, Optimisers.Leaf{Adam{Float32}, Tuple{CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, Tuple{Float32, Float32}}}}}, x::ComponentVector{Float32, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, Tuple{Axis{(a = 1:2, b = 3:4)}}}, x̄s::ComponentVector{Float32, CuAArray{Float32, 1, CUDA.Mem.DeviceBuffer}, Tuple{Axis{(a = 1:2, b = 3:4)}}})
    @ Optimisers C:\Users\Luffy\.julia\packages\Optimisers\xUXSc\src\interface.jl:43
 [14] update(tree::NamedTuple{(:a, :b), Tuple{Optimisers.Leaf{Adam{Float32}, Tuple{CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, Tuple{Float32, Float32}}}, Optimisers.Leaf{Adam{Float32}, Tuple{CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, Tuple{Float32, Float32}}}}}, x::ComponentVector{Float32, CuArray{Float32, 1, CUDA.Mem.DeviceBuffer}, Tuple{Axis{(a = 1:2, b = 3:4)}}}, x̄s::ComponentVector{Float32, CuArrray{Float32, 1, CUDA.Mem.DeviceBuffer}, Tuple{Axis{(a = 1:2, b = 3:4)}}})
    @ Optimisers C:\Users\Luffy\.julia\packages\Optimisers\xUXSc\src\interface.jl:49
 [15] top-level scope
    @ REPL[25]:1
```
Unless you want me to recover these overloadings 
https://github.com/avik-pal/Lux.jl/pull/109/files#diff-47c27891e951c8cd946b850dc2df31082624afdf57446c21cb6992f5f4b74aa2L159

Reply:
Mainly because ComponentArray is a structured array and we might want to apply operations exclusively on a leaf with fmap similar to what we do for NamedTuples.

Reply:
I understand your point, but the current version would give us just one leaf
```julia
julia> ps = ComponentArray(cu[1f0,2f0,3f0,4f0], Axis(a=1:2, b=3:4));

julia> st = Optimisers.setup(Adam(), ps)
Leaf(Adam{Float32}(0.001, (0.9, 0.999), 1.19209f-7), (Float32[0.0, 0.0, 0.0, 0.0], Float32[0.0, 0.0, 0.0, 0.0], (0.9, 0.999)))
```
I don't know how to achieve what you said in this PR.

Reply:
Is this with the optimisers patch or without it?

Reply:
The release version, so with those patches. I think the reason is that we cannot create a `ComponentArray` from a named tuple composite of `CuArray`s
```julia
julia> ps = ComponentArray(a = cu[1,2], b = cu[3,4]);
ERROR: CuArray only supports element types that are stored inline
Stacktrace:
  [1] error(s::String)
    @ Base .\error.jl:33
  [2] CuArray{Vector{Any}, 1, CUDA.Mem.DeviceBuffer}(#unused#::UndefInitializer, dims::Tuple{Int64})
    @ CUDA C:\Users\Luffy\.julia\packages\CUDA\DfvRa\src\array.jl:34
  [3] (CuArray{Vector{Any}, 1})(#unused#::UndefInitializer, dims::Tuple{Int64})
    @ CUDA C:\Users\Luffy\.julia\packages\CUDA\DfvRa\src\array.jl:125
  [4] (CuArray{Vector{Any}})(#unused#::UndefInitializer, dims::Tuple{Int64})
    @ CUDA C:\Users\Luffy\.julia\packages\CUDA\DfvRa\src\array.jl:136
  [5] similar(bc::Base.Broadcast.Broadcasted{CUDA.CuArrayStyle{1}, Tuple{Base.OneTo{Int64}}, typeof(identity), Tuple{CuArray{Int64, 1, CUDA.Mem.DeviceBuffer}}}, #unused#::Type{Vector{Any}}, dims::Tuple{Int64})
    @ CUDA C:\Users\Luffy\.julia\packages\CUDA\DfvRa\src\broadcast.jl:14
  [6] _mapreduce(f::typeof(identity), op::ComponentArrays.var"#28#29", As::CuArray{Int64, 1, CUDA.Mem.DeviceBuffer}; dims::Colon, init::Vector{Any})
    @ GPUArrays C:\Users\Luffy\.julia\packages\GPUArrays\Hyss4\src\host\mapreduce.jl:64
  [7] mapreduce(::Function, ::Function, ::CuArray{Int64, 1, CUDA.Mem.DeviceBuffer}; dims::Function, init::Vector{Any})
    @ GPUArrays C:\Users\Luffy\.julia\packages\GPUArrays\Hyss4\src\host\mapreduce.jl:31
  [8] reduce(op::Function, A::CuArray{Int64, 1, CUDA.Mem.DeviceBuffer}; kw::Base.Pairs{Symbol, Vector{Any}, Tuple{Symbol}, NamedTuple{(:init,), Tuple{Vector{Any}}}})
    @ Base .\reducedim.jl:371
  [9] pushcat!(a::Vector{Any}, b::CuArray{Int64, 1, CUDA.Mem.DeviceBuffer})
    @ ComponentArrays C:\Users\Luffy\.julia\packages\ComponentArrays\Eoni9\src\componentarray.jl:223
 [10] make_idx(data::Vector{Any}, x::CuArray{Int64, 1, CUDA.Mem.DeviceBuffer}, last_val::Int64)
    @ ComponentArrays C:\Users\Luffy\.julia\packages\ComponentArrays\Eoni9\src\componentarray.jl:178
 [11] make_idx(data::Vector{Any}, nt::NamedTuple{(:a, :b), Tuple{CuArray{Int64, 1, CUDA.Mem.DeviceBuffer}, CuArray{Int64, 1, CUDA.Mem.DeviceBuffer}}}, last_val::Int64)
    @ ComponentArrays C:\Users\Luffy\.julia\packages\ComponentArrays\Eoni9\src\componentarray.jl:156
 [12] make_carray_args(A::Type{Vector}, nt::NamedTuple{(:a, :b), Tuple{CuArray{Int64, 1, CUDA.Mem.DeviceBuffer}, CuArray{Int64, 1, CUDA.Mem.DeviceBuffer}}})
    @ ComponentArrays C:\Users\Luffy\.julia\packages\ComponentArrays\Eoni9\src\componentarray.jl:146
 [13] make_carray_args(nt::NamedTuple{(:a, :b), Tuple{CuArray{Int64, 1, CUDA.Mem.DeviceBuffer}, CuArray{Int64, 1, CUDA.Mem.DeviceBuffer}}})
    @ Lux C:\Users\Luffy\.julia\packages\Lux\eYTuG\src\utils.jl:144
 [14] ComponentArray(nt::NamedTuple{(:a, :b), Tuple{CuArray{Int64, 1, CUDA.Mem.DeviceBuffer}, CuArray{Int64, 1, CUDA.Mem.DeviceBuffer}}})
    @ ComponentArrays C:\Users\Luffy\.julia\packages\ComponentArrays\Eoni9\src\componentarray.jl:63
 [15] ComponentArray(; kwargs::Base.Pairs{Symbol, CuArray{Int64, 1, CUDA.Mem.DeviceBuffer}, Tuple{Symbol, Symbol}, NamedTuple{(:a, :b), Tuple{CuArray{Int64, 1, CUDA.Mem.DeviceBuffer}, CuArray{Int64, 1, CUDA.Mem.DeviceBuffer}}}})
    @ ComponentArrays C:\Users\Luffy\.julia\packages\ComponentArrays\Eoni9\src\componentarray.jl:67
 [16] top-level scope
    @ REPL[18]:1
 [17] top-level scope
    @ C:\Users\Luffy\.julia\packages\CUDA\DfvRa\src\initialization.jl:52
```

Reply:
This is something that I already know. But what does it have to do with `fmap`ping on a  component array?